File: app/src/main/java/deltazero/amarok/FileHider/ChmodFileHider.java
package deltazero.amarok.FileHider;

import android.content.Context;
import android.util.Log;

import java.util.HashSet;
import java.util.Set;

import deltazero.amarok.utils.ShellUtil;

public class ChmodFileHider extends BaseFileHider {

    public ChmodFileHider(Context context) {
        super(context);
    }

    @Override
    protected void process(Set<String> targetDirs, ProcessMethod method) throws InterruptedException {

        var processDirs = new HashSet<String>();
        for (var d : targetDirs) {
            if (d.startsWith("/storage/emulated/0/"))
                processDirs.add(d.replace("/storage/emulated/", "/data/media/"));
            else
                Log.w("ChmodFileHider", String.format("Unsupported path: %s", d));
        }

        StringBuilder cmd = new StringBuilder("echo amarok");
        for (String d : processDirs)
            cmd.append(String.format("&chmod -R %s %s", method == ProcessMethod.HIDE ? 0 : 2770, d));
        ShellUtil.exec(new String[]{"su", "-c", cmd.toString()});
    }
}

File: app/src/main/java/deltazero/amarok/filehider/ChmodFileHider.kt
package deltazero.amarok.filehider

import android.content.Context
import android.util.Log
import com.topjohnwu.superuser.Shell
import deltazero.amarok.R
import deltazero.amarok.core.ActivationResult
import deltazero.amarok.core.HideAction
import deltazero.amarok.utils.MediaStoreHelper
import deltazero.amarok.utils.await
import deltazero.amarok.utils.awaitShell

class ChmodFileHider(private val context: Context) : FileHider {
  override val mode = FileHiderMode.CHMOD
  override val name = "Chmod"

  override suspend fun activate(): ActivationResult {
    val shell = awaitShell()
    return if (shell.isRoot) {
      ActivationResult(success = true, msgResId = 0)
    } else {
      ActivationResult(success = false, msgResId = R.string.root_not_ava)
    }
  }

  override suspend fun process(targetDirs: Set<String>, action: HideAction) {
    val hide = action is HideAction.Hide
    val processDirs = mutableSetOf<String>()
    for (d in targetDirs) {
      if (d.startsWith("/storage/emulated/")) {
        processDirs.add(d.replace("/storage/emulated/", "/data/media/"))
      } else {
        Log.w("ChmodFileHider", "Unsupported path: $d")
      }
    }

    val perm = if (hide) "0" else "2770"
    val processJob = Shell.getShell().newJob()
    for (d in processDirs) {
      processJob.add("chmod -R $perm $d")
    }
    processJob.await()
    MediaStoreHelper.scan(context, processDirs)
  }
}
